File: Examples/ObjectiveC/ExtrusionsExample.m
//
//  ExtrusionsExample.m
//  Examples
//
//  Created by Jordan Kiley on 5/11/17.
//  Copyright © 2017 Mapbox. All rights reserved.
//

#import "ExtrusionsExample.h"

@import Mapbox;

NSString *const MBXExample3DExtrusions = @"ExtrusionsExample";

@interface ExtrusionsExample () <MGLMapViewDelegate>

@end

@implementation ExtrusionsExample

- (void)viewDidLoad {
    [super viewDidLoad];
    
    MGLMapView *mapView = [[MGLMapView alloc] initWithFrame:self.view.bounds styleURL:[MGLStyle lightStyleURLWithVersion:9]];
    
    // Center the map on the Colosseum in Rome, Italy.
    [mapView setCenterCoordinate:CLLocationCoordinate2DMake(41.8902, 12.4922)];
    
    // Set the map view camera's pitch and distance.
    mapView.camera = [MGLMapCamera cameraLookingAtCenterCoordinate:mapView.centerCoordinate fromDistance:600 pitch:60 heading:0];
    mapView.delegate = self;
    
    [self.view addSubview:mapView];
}

- (void)mapView:(MGLMapView *)mapView didFinishLoadingStyle:(MGLStyle *)style {
    
    // Access the Mapbox Streets source and use it to create a `MGLFillExtrusionStyleLayer`. The source identifier is `composite`. Use the `sources` property on a style to verify source identifiers.
    MGLSource *source = [style sourceWithIdentifier:@"composite"];
    MGLFillExtrusionStyleLayer *layer = [[MGLFillExtrusionStyleLayer alloc] initWithIdentifier:@"buildings" source:source];
    layer.sourceLayerIdentifier = @"building";
    
    // Filter out buildings that should not extrude.
    layer.predicate = [NSPredicate predicateWithFormat:@"extrude != false AND height >=0"];
    
    // Set the fill extrusion height to the value for the building height attribute.
    layer.fillExtrusionHeight = [MGLStyleValue valueWithInterpolationMode:MGLInterpolationModeIdentity sourceStops:nil attributeName:@"height" options:nil];
    layer.fillExtrusionOpacity = [MGLStyleValue valueWithRawValue:@0.75];
    layer.fillExtrusionColor = [MGLStyleValue valueWithRawValue:[UIColor whiteColor]];
    
    // Insert the fill extrusion layer below a POI label layer. Use the `layers` property on a style to verify built-in layer identifiers.
    MGLStyleLayer *symbolLayer = [style layerWithIdentifier:@"poi-scalerank3"];
    [style insertLayer:layer belowLayer:symbolLayer];
}

@end

File: Examples/Swift/ExtrusionsExample.swift
//
//  ExtrusionsExample.swift
//  Examples
//
//  Created by Jordan Kiley on 5/10/17.
//  Copyright © 2017 Mapbox. All rights reserved.
//

import Mapbox

@objc(ExtrusionsExample_Swift)

class ExtrusionsExample: UIViewController, MGLMapViewDelegate {
    
    var mapView : MGLMapView!
    override func viewDidLoad() {
        super.viewDidLoad()
        
        mapView = MGLMapView(frame: view.bounds, styleURL: MGLStyle.lightStyleURL(withVersion: 9))
        mapView.setCenter(CLLocationCoordinate2D(latitude: 41.8902, longitude: 12.4922), animated: false)

        // Set the `MGLMapCamera` to one
        mapView.camera = MGLMapCamera(lookingAtCenter: mapView.centerCoordinate, fromDistance: 600, pitch: 60, heading: 0)

        mapView.delegate = self
        view.addSubview(mapView)
    }
    
    func mapView(_ mapView: MGLMapView, didFinishLoading style: MGLStyle) {
        
        // Access the Mapbox Streets source.
        let source = style.source(withIdentifier: "composite")
        
        // Create a `MGLFillExtrusionLayer` using the building layer from that source.
        let layer = MGLFillExtrusionStyleLayer(identifier: "buildings", source: source!)
        layer.sourceLayerIdentifier = "building"
        
        // Check that the building is should
        layer.predicate = NSPredicate(format: "extrude != false AND height >=0")
        
        layer.fillExtrusionHeight = MGLStyleValue(interpolationMode: .identity, sourceStops: nil, attributeName: "height", options: nil)
        
        layer.fillExtrusionOpacity = MGLStyleValue(rawValue: 0.75)
        layer.fillExtrusionColor = MGLStyleValue(rawValue: .white)
        let symbolLayer = style.layer(withIdentifier: "poi-scalerank3")
        style.insertLayer(layer, below: symbolLayer!)
        
    }
    
}
